### PR TITLE
Improving Node Discovery via DHT Announcements - fixes  #152

### LINKED ISSUE #152: Do topic discovery for the pubsub group using a DHT
Currently, we have a big pubsub group where everyone sends messages announcing themselves to discover other peers and the pubsub topics. This doesn't scale.

We want to change it for a DHT that have a `Map<topic_id, nodes[]>`

### PR DESCRIPTION
**Pull request by elielnfinic**
### PR Overview:

This pull request addresses issue #152, focusing on enhancing node discovery for peers subscribed to specific topics by utilizing the DHT (Distributed Hash Table) rather than relying on a centralized PubSub mechanism.

### Summary of Changes:
DHT Integration:  Enabled DHT on the node, allowing topics to map directly to the peers subscribed to them.
Updated `packages/network/node.ts` file:

Added methods: `putDataOnDHT()getDataFromDHT(), announcePeerOnDHT(), removePeerFromDHT(), getPeersOnTopicFromDHT()`

Created a new private property:  `_dht`

Added the following external libraries to the network package:
	•	@libp2p/kad-dht
	•	@libp2p/peer-id
	•	@it-last

### CODE REVIEW

File: packages/network/package.json
Comment (d-roak) [inline packages/network/package.json:1]: don't change indentation

File: packages/network/src/node.ts [hunk 1]
```diff
@@ -15,9 +15,13 @@ import { generateKeyPairFromSeed } from "@libp2p/crypto/keys";
 import { dcutr } from "@libp2p/dcutr";
 import { devToolsMetrics } from "@libp2p/devtools-metrics";
 import { identify } from "@libp2p/identify";
+
 import type {
 	Ed25519PeerId,
 	EventCallback,
+	EventHandler,
+	PeerId,
+	PeerInfo,
```
Comment (d-roak) [inline packages/network/src/node.ts:24]: not used

File: packages/network/src/node.ts [hunk 2]
```diff
@@ -15,9 +15,13 @@ import { generateKeyPairFromSeed } from "@libp2p/crypto/keys";
 import { dcutr } from "@libp2p/dcutr";
 import { devToolsMetrics } from "@libp2p/devtools-metrics";
 import { identify } from "@libp2p/identify";
+
 import type {
 	Ed25519PeerId,
 	EventCallback,
+	EventHandler,
```
Comment (d-roak) [inline packages/network/src/node.ts:22]: not used

File: packages/network/src/node.ts [hunk 3]
```diff
@@ -26,13 +30,22 @@ import type {
 	StreamHandler,
 	URLPeerId,
 } from "@libp2p/interface";
+
+import {
+	type KadDHT,
+	type ValueEvent,
+	kadDHT,
+	removePublicAddressesMapper,
```
Comment (d-roak) [inline packages/network/src/node.ts:38]: not used

File: packages/network/src/node.ts [hunk 4]
```diff
@@ -26,13 +30,22 @@ import type {
...
+} from "@libp2p/kad-dht";
 import { peerIdFromString } from "@libp2p/peer-id";
 import { pubsubPeerDiscovery } from "@libp2p/pubsub-peer-discovery";
 import { webRTC, webRTCDirect } from "@libp2p/webrtc";
 import { webSockets } from "@libp2p/websockets";
 import { webTransport } from "@libp2p/webtransport";
 import { multiaddr } from "@multiformats/multiaddr";
+import last from "it-last";
 import { type Libp2p, createLibp2p } from "libp2p";
+import { toString as uint8ToString } from "uint8arrays";
```
Comment (d-roak) [inline packages/network/src/node.ts:48]: keep consistency `uint8ArrayToString`

File: packages/network/src/node.ts [hunk 5]
```diff
@@ -97,6 +111,7 @@ export class TopologyNetworkNode {
 				autonat: autoNAT(),
 				dcutr: dcutr(),
 				identify: identify(),
+				dht: kadDHT({}),
```
Comment (d-roak) [inline packages/network/src/node.ts:114]: keep consistency, when not passing configs, don't pass anything as an input

File: packages/network/src/node.ts [hunk 6]
```diff
@@ -248,4 +274,103 @@ export class TopologyNetworkNode {
...
+	}
+
+	/*
+	 * Get the peers on a topic from the DHT
+	 * @param topic The topic to get the peers from
+	 * @returns A set of PeerId
+	 * */
+	async getPeersOnTopicFromDHT(topic: string): Promise<Set<PeerId>> {
+		const uint8Topic = uint8ArrayFromString(topic);
+		const peersOnTopic = await this._dht?.get(uint8Topic);
```
Comment (d-roak) [inline packages/network/src/node.ts:364]: no need for `await`

File: packages/network/src/node.ts [hunk 7]
```diff
@@ -248,4 +274,103 @@ export class TopologyNetworkNode {
...
+			return false;
+		}
+	}
+
+	/**
+	 *
+	 * @param key The key to search for
+	 * @returns The value if the data was found, undefined if the data was not found or the DHT is not initialized
+	 */
+	async getDataFromDHT(key: Uint8Array): Promise<Uint8Array | undefined> {
```
Comment (d-roak) [inline packages/network/src/node.ts:308]: why do we need this method since we have `getPeersOnTopicFromDHT` ?

File: packages/network/src/node.ts [hunk 8]
```diff
@@ -151,6 +167,14 @@ export class TopologyNetworkNode {
 		try {
 			this._pubsub?.subscribe(topic);
 			this._pubsub?.getPeers();
+			this.anouncePeerOnDHT(topic, this._node.peerId);
+
+			// connect to all peers on the topic
+			const peers = await this.getPeersOnTopicFromDHT(topic);
+			for (const peerId of peers) {
+				await this._node.dial(peerId);
```
Comment (d-roak) [inline packages/network/src/node.ts:175]: can't do this. put a `topic_discovery_peers_threshold` variable in the network config and use that threshold as the minimum number of peers.
If it falls down the threshold we try to connect, if it's equal we skip (this is the same everywhere, might segment this logic)

File: packages/network/src/node.ts [hunk 9]
```diff
@@ -248,4 +274,103 @@ export class TopologyNetworkNode {
...
+	 * @param value  The value to search for
+	 * @returns The value `true` if the data was put on the DHT successfully, `false` if not and undefined if the DHT is not initialized
+	 */
+	async putDataOnDHT(
+		key: Uint8Array,
+		value: Uint8Array,
+	): Promise<boolean | undefined> {
+		if (!this._dht) {
+			console.error("DHT not initialized. Please run .start()");
+			return undefined;
```
Comment (d-roak) [inline packages/network/src/node.ts:290]: return false here

File: packages/network/src/node.ts [hunk 10]
```diff
@@ -248,4 +274,103 @@ export class TopologyNetworkNode {
...
+		}
+
+		try {
+			await this._dht?.put(key, value);
+			console.log("Successfully stored data in DHT");
```
Comment (d-roak) [inline packages/network/src/node.ts:295]: change this log to the same format as the other logs

**Review by d-roak — CHANGES_REQUESTED** [review]
partial review

File: packages/network/src/node.ts (hunk 6 above)
Comment (elielnfinic) [reply packages/network/src/node.ts]: You are right, I have just tested it.

File: packages/network/src/node.ts (hunk 8 above)
Comment (elielnfinic) [reply packages/network/src/node.ts]: Added a test to make sure I can't dial myself.

File: packages/network/src/node.ts (hunk 7 above)
Comment (elielnfinic) [reply packages/network/src/node.ts]: I have removed it. Thank you for pointing that out.

File: packages/network/package.json
Comment (d-roak) [reply packages/network/package.json:1]: not resolved

Comment (elielnfinic) [reply packages/network/package.json:1]: I have resolved it.

**Pull request closed without merging**